### PR TITLE
Add search scopes

### LINKED ISSUE #37: Add help text below the search field
Let people know that they can search by Product Name, E number, …

### PR DESCRIPTION
**Pull request by bash**
Resolves  #37

## Preview
Screenshot 2019-06-09 at 11 34 22

## Todo
- [x] Send scope to backend
- [x] Respect search scope in backend
- [x] Allow e numbers to be typed without the leading "E"

### CODE REVIEW

**Comment by bash** [comment]
Screenshot 2019-07-21 at 21 27 33

File: client/src/components/search-input.tsx [hunk 1]
```diff
@@ -1,19 +1,35 @@
...
 }
+
+const mapToHtmlInputMode = (type: SearchInputType): string => {
+ switch (type) {
+   case SearchInputType.Default:
+     return 'search';
+   case SearchInputType.NumbersOnly:
+     return 'numeric';
+ }
+}
```
Comment (jnferner) [inline client/src/components/search-input.tsx]: Terminating newline is missing. Is there an error in our `.editorconfig`? Do we even have one? 🤔

File: client/src/components/search-results.tsx [hunk 2]
```diff
@@ -47,25 +51,25 @@ export class SearchResults extends Component<SearchResultsProps, SearchResultsSt
   }
 
   componentWillReceiveProps(nextProps: SearchResultsProps, _: SearchResultsState) {
-    if (this.state.query !== nextProps.query) {
-      this.fetchItems(nextProps.query);
+    if (this.state.query !== nextProps.query || this.state.scope !== nextProps.scope) {
```
Comment (jnferner) [inline client/src/components/search-results.tsx]:
```suggestion
    if (!(this.state.query === nextProps.query && this.state.scope === nextProps.scope)) {
```

File: client/src/components/search-scope-bar.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    </label>
+  )
+}
+
+function labelForScope(scope: SearchScope): string {
+  switch (scope) {
+    case SearchScope.Names:
+      return 'Search names';
+    case SearchScope.ENumbers:
+      return 'Search E\u{00A0}numbers';
```
Comment (jnferner) [inline client/src/components/search-scope-bar.tsx]: Can we put `\u{00A0}` in a constant?

File: scripts/setup_vm.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+EOF
+
+dnf install -y elasticsearch java-openjdk-devel java-openjdk
+
+echo 'network.host: 0.0.0.0' >> /etc/elasticsearch/elasticsearch.yml
+
+systemctl enable --now elasticsearch
+
+firewall-cmd --zone=public --add-port=9200/tcp --permanent
+firewall-cmd --reload
```
Comment (jnferner) [inline scripts/setup_vm.sh]: Terminating newline is missing

File: src/server.rs [hunk 5]
```diff
@@ -47,13 +49,41 @@ impl Server for RocketServer {
...
+            _ => Err(form_value),
+        }
+    }
+}
+
+/// Searches a single item with a scope
+#[get("/search?<query>&<scope>")]
 fn search(
     query: String,
+    scope: Option<Scope>,
```
Comment (jnferner) [inline src/server.rs]: Are you sure we should allow omitting the scope?

File: client/src/components/search-scope-bar.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        return <Item scope={scope} onSelect={() => onSelectScope(scope)} selected={scope === selectedScope} />
+      })}
+    </div>
+  )
+}
+
+const Item = ({ scope, selected, onSelect }: { scope: SearchScope, selected: boolean, onSelect: () => void }) => {
+  return (
+    <label class='item'>
+      <input type='radio' value={scope} name='search-scope' class='radio' onInput={onSelect} checked={selected} />
```
Comment (jnferner) [inline client/src/components/search-scope-bar.tsx]: Can we split this on multiple lines?

File: client/src/components/search-input.tsx (hunk 1 above)
Comment (bash) [reply client/src/components/search-input.tsx]: We don't have one 🤷‍♂

Comment (bash) [reply client/src/components/search-input.tsx]: d05dfd1

File: scripts/setup_vm.sh (hunk 4 above)
Comment (bash) [reply scripts/setup_vm.sh]: d05dfd1

File: client/src/components/search-results.tsx (hunk 2 above)
Comment (bash) [reply client/src/components/search-results.tsx]: 072e56a

File: client/src/components/search-scope-bar.tsx (hunk 3 above)
Comment (bash) [reply client/src/components/search-scope-bar.tsx]: That's obviously a … *(googles)* … "no-break space". duh. 😆

Comment (bash) [reply client/src/components/search-scope-bar.tsx]: 9f136a0

File: client/src/components/search-scope-bar.tsx (hunk 6 above)
Comment (bash) [reply client/src/components/search-scope-bar.tsx]: b9f930a

File: src/server.rs (hunk 5 above)
Comment (bash) [reply src/server.rs]: 67b5841

**Pull request merged**